### PR TITLE
[OMPIRBuilder] - Handle dependencies in `createTarget`

### PR DESCRIPTION
**Pull request by bhandarkar-pranav**
This PR handles dependencies specified by the `depend` clause on an OpenMP target construct. It does this much the same way clang does it by materializing an OpenMP `task` that is tagged with the dependencies.

To aid review, here are the relevant implementation details
1) `createTarget` - This function itself is largely unchanged except that it now accepts a vector of `DependData` objects that it simply forwards to `emitTargetCall`
2) `emitTargetCall` - This function has changed now to check if an outer target-task needs to be materialized (Yes, if target construct has `nowait` or has `depend` clause). If yes, it calls `emitTargetTask` to do all the heavy lifting for creating and dispatching the task.
3) `emitTargetTask` - Bulk of the PR is here. See the large comment explaining what it does at the beginning of the function.

Things that will/should follow in subsequent PRs
1) Handle this for target enter data, exit data and update data constructs
2) (Nice to have) - Outline logic for creating code for a task and use in `createTask` and `emitTargetTask`

### CODE REVIEW

**Comment by bhandarkar-pranav** [comment]
ping! Gentle reminder, I'd appreciate reviews on this.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (ergawy) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5372]: Really appreciate this block of comments. Paints the picture of what happens clearly. Thanks :)!

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (ergawy) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5491]: I there are pieces in this lambda that can be outlined into shared utils and called from here and `OpenMPIRBuilder::createTask`. For example, the logic to setup `TaskData` (but, other parts as well).

**Review by ergawy — COMMENTED** [review]
Thanks @pranavb-ca! I did a first round and have a few comments.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (skatrak) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6003]: Nit: Is this addition intentional?

**Review by skatrak — COMMENTED** [review]
Thank you Pranav, I think this looks good. I just have a few minor comments.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:797]: Sorry for having left these here. This should go.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:1710]: Good point. I'll rename this function.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:1721]: Aah nice catch.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:1730]: Will do, thanks.

Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:1730]: Actually, my intention was to switch `OpenMPIRBuilder::createTask` to use this function but i had a couple of other changes to `OpenMPIRBuilder::createTask` in mind so I decided to leave that out and roll it into a different PR. If you prefer, I can make the switch in this PR itself though. Just a preference, no strong opinions on this.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5281]: Good point @ergawy and @skatrak I'll pick a different name `emitTargetTaskProxyFunction` sounds appropriate.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5288]: Yes, that's a better name.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5306]: Got it,  yes. will do that.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5321]: Yes, this should have been removed before opening the PR even. I accidentally left something I had put in place when i was debugging my work.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5344]: Sure, will do.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5664]: No, this should go. sorry about this.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6003]: No, this should go :(

File: offload/test/offloading/fortran/target-depend.f90
Comment (bhandarkar-pranav) [reply offload/test/offloading/fortran/target-depend.f90:26]: Yes, that makes sense. I'll add a omp parallel also so that if I remove the depend there is a chance that i'd get the wrong result. That'll really prove this works.

File: llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h
Comment (Meinersbur) [inline llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:2279]:
```suggestion
                             ArrayRef<DependData> Dependencies = {});
```

File: llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h
Comment (Meinersbur) [inline llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:1783]:
```suggestion
      ArrayRef<OpenMPIRBuilder::DependData> Dependencies, bool HasNoWait);
```

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (Meinersbur) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:1713]:
```suggestion
  if (Dependencies.empty())
```

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (Meinersbur) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:1724]:
```suggestion
    Builder.SetInsertPoint(
        &OldIP.getBlock()->getParent()->getEntryBlock().getTerminator());
```
just makes it more explicit. 
Did you consider explicitly passing the AllocaIP, could be needed with combined constructs.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (Meinersbur) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:1754]: Consider adding a comment with a mock source of what is generate here:
```
DepArray[P].BaseAddre = ...
```

**Review by Meinersbur — COMMENTED** [review]
Not a full a review, but the first notes that I started with

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (Meinersbur) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:684]: [nit]
```suggestion

static void
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (Meinersbur) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:687]:
```suggestion
                SmallVectorImpl<llvm::OpenMPIRBuilder::DependData> &dds) {
```

**Review by Meinersbur — COMMENTED** [review]
Conceptually, looks quite good. Just some style comments.

File: llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h
Comment (bhandarkar-pranav) [reply llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:1783]: I don't think this is safe. I capture Dependencies in a lambda so storing ArrayRef which doesnt manage the lifetime of the data pointer isn't safe.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5491]: Yes, that is an active Todo which I was planning to add in a different PR because now this PR doesn't touch the handling of the `task` construct at all. But that's only a preference, I'll be happy to roll those changes into this PR too.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5335]: Thanks

**Review by ergawy — APPROVED** [review]
LGTM

**Comment by bhandarkar-pranav** [comment]
@Meinersbur, @ergawy, @skatrak  - Could you please give this another look. I have updated the testcase `target-depend.f90` because the previous version was flaky; I had used `depend` incorrectly which I have fixed now. (Previously, `depend` was used on tasks that were not sibling tasks)

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5709]: Thanks for point this out. I want to use this opportunity to understand this idiom I see all around in `OMPIRBuilder.cpp` which is ``Builder.restoreIP(FunctThatReturnsInsertPointTy())``. Now, if the only return from `FuncThatReturnsInsertPointTy` is `Builder.saveIP()`, then do I need a call to `Builder.restoreIP()` on return ? Because `emitTargetTask()` returns `Builder.saveIP()`, which, in other words, means it is returning the state of the Builder's insert point. We then restore the same `Builder` to that insert point which sounds redundant. However, I see this in a number of places in `OMPIRBuilder.cpp`

Comment (skatrak) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5709]: I agree with you, since I have noticed this redundant pattern as well, though in a way I think it makes sense from a uniformity point of view. There are several codegen functions that return an insertion point, which should be used by the caller to proceed with any further codegen.

I think that the fact that oftentimes the insertion point that's being returned is indeed already the current one shouldn't be relied on by the caller unless we make this a contract followed by all such functions. To be consistent, I think these codegen function must either always return an insertion point which must be used by the caller (the current approach) or all codegen functions must not return an insertion point and set it themselves to the spot follow-up codegen should go.

In my opinion, I believe it's currently better to keep the current approach even though it results in some redundant setting of the insertion point. Skipping the call to `Builder.restoreIP()` in the caller could result in unintended problems in the future, since the current contract is that the returned insertion point is the one that should be used and not the current one.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:732]: It looks like this comment hasn't been addressed. Let me know if you just missed it, if you disagree with it or if it isn't very clear and I should try to explain better.

**Review by skatrak — COMMENTED** [review]
Thanks for working on my previous comments. I just have a couple more minor ones.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5521]: This one is intentional because as more work is done for `target task` (and more work remains) being able to see this function will be useful for debugging. At a later, more stable point, we can get rid of this. Just my preference to leave this around for now.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (bhandarkar-pranav) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3106]: @skatrak  - Apologies for missing your previous comment about moving the empty depends list check into `buildDependData`. I have done that now in this iteration of the PR.

**Comment by bhandarkar-pranav** [comment]
@skatrak - I have updated the PR with changes based on your feedback. Could you please take a look?

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:686]: I'd suggest taking `std::optional<ArrayAttr> depends` and `OperandRange dependVars` as arguments instead of `Operation *op` here. This way you don't need to try casting `op` to each supported operation below, but rather make the body of this function what the `processDepends` lambda inside currently has.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:743]: Suggestion to how this call could look if following my comment above:
```suggestion
  buildDependData(moduleTranslation, taskOp.getDepends(), taskOp.getDependVars(), dds);
```

**Review by skatrak — APPROVED** [review]
Thank you Pranav, LGTM. I have just one more comment but no need to wait for another review by me before merging this. Just make sure @Meinersbur is happy with it as well!

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5493]: @Meinersbur  -> The only reason, I didnt do it is because `createFakeIntVal` is used by a few other functions not related to this PR. So changing `ToBeDeleted` from `std::stack`  to `SmallVector` would involve changing other functions (such as `createTeams`) which are not related to this PR. I could go ahead and add that to this PR. Or simply open a separate PR just to make this change which should be quick. Let me know what you prefer.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (Meinersbur) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5510]: This passes `ToBeDeleted` and `Dependencies` by-value to the lambda, i.e. makes a copy of those lists. If called multiple times, items in `ToBeDeleted` will be erased multiple times. I assume this is not intended?
```suggestion
  OI.PostOutlineCB = [this, &ToBeDeleted, &Dependencies,
```

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp]: @Meinersbur - Please take a look. I have gone ahead and made this change. If you agree with my reason above then I'll simply back out that one [commit](https://github.com/llvm/llvm-project/pull/93977/commits/a314a84964ae243cb4fee98a9adeacdfa9edf71e) and make it a new PR.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (bhandarkar-pranav) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5508]: > This passes `ToBeDeleted` and `Dependencies` by-value to the lambda, i.e. makes a copy of those lists. If called multiple times, items in `ToBeDeleted` will be erased multiple times. I assume this is not intended?

In fact a copy is necessary to "own" the data. `PostOutlineCB` is called only once so multiple deletions is not a problem. However, and what is more important to note, is that it is called long after `ToBeDeleted` is destroyed (that is after we have returned from `emitTargetTask`) making it necessary to keep a copy around.

**Review by Meinersbur — APPROVED** [review]
LGTM

**Pull request merged**
